### ISSUE TITLE
[Performance] Add detailed GPU kernel timing for bottleneck identification

### ISSUE THREAD

**Issue by dpalmqvist** [motsognirr/olmlx#227]
## Description

Currently there's no detailed breakdown of where time is spent during inference (attention vs FFN vs logits vs memory transfer). This makes it difficult to optimize specific bottlenecks for different model sizes.

## Current State

Only high-level timing:
- `load_duration` - model loading time
- `prompt_eval_count` / `eval_count` - token counts
- `prompt_tps` / `gen_tps` - tokens per second

## Suggested Approach

1. Add per-operation timing in `inference.py`:
   - Attention forward time
   - FFN forward time  
   - Logits computation time
   - KV cache update time
   - Memory I/O time (for Flash models)

2. Add to timing stats:
   ```python
   @dataclasses.dataclass
   class TimingStats:
       # Existing
       load_duration: int = 0
       eval_duration: int = 0
       total_duration: int = 0
       
       # New
       attention_duration: int = 0
       ffn_duration: int = 0
       logits_duration: int = 0
       kv_cache_duration: int = 0
       io_duration: int = 0  # for Flash SSD reads
   ```

3. Add logging of breakdown on generation complete

## Files to Modify

- `olmlx/utils/timing.py` - extend TimingStats
- `olmlx/engine/inference.py` - add timing hooks
- `olmlx/engine/flash/flash_mlp.py` - add I/O timing for Flash

## Priority

Medium - enables data-driven optimization decisions

Related: olmlx/engine/inference.py:1996-2004

**Comment by dpalmqvist** [comment]
## Analysis before implementation

A few design concerns with the approach as sketched — worth resolving before anyone picks this up.

### 1. MLX is lazy — naive `Timer` wrapping won't measure compute

Wrapping `attn_output = self.attention(...)` in a `with Timer():` block records essentially zero time, because MLX only builds a graph at that point. Real compute happens later at the next `mx.eval()` / materialization.

To get honest per-op numbers you'd need to force `mx.eval(...)` at every timing boundary. That:
- serializes the pipeline and destroys kernel fusion,
- meaningfully slows inference when enabled,
- makes the numbers only comparable to each other under the same forced-sync regime.

### 2. The timing points live in the wrong layer

Attention / FFN / logits / KV-cache update happen **inside `mlx-lm`'s model forward**, not inside `olmlx/engine/inference.py`. From `inference.py`'s perspective the whole step is one `model(x)` call.

Instrumenting these requires either:
- wrapping/patching upstream transformer blocks (attention, MLP, lm_head) at model-load time, or
- using MLX's Metal profiling hooks (`mx.metal.start_capture` / `stop_capture`) to get kernel-level traces in Xcode Instruments.

Just extending `TimingStats` in `utils/timing.py` is not sufficient — there has to be a hook point inside the forward pass.

### 3. API pollution

Adding `attention_duration`, `ffn_duration`, etc. to `TimingStats.to_dict()` puts them into Ollama `/api/generate` and `/api/chat` responses. That's a public API surface change for numbers that:
- only exist when forced-sync instrumentation is on,
- are misleading without that instrumentation (they'd reflect graph-build time, not compute).

Diagnostic timing should not ride on the response schema.

### Suggested directions

**A. Opt-in profiling mode** (preferred)
- Env flag, e.g. `OLMLX_PROFILE=1`.
- When set, wrap the loaded model's transformer blocks (attention, MLP, lm_head) with timing shims that `mx.eval()` at boundaries and accumulate per-category ns counters.
- Log the breakdown to the server log on generation complete.
- Zero overhead when disabled. Response schema untouched.
- Probably belongs in `engine/model_manager.py` (wrap on load) + a new `engine/profiling.py` module, with accumulators exposed to `inference.py` for end-of-request logging.

**B. Metal capture hook** (complementary)
- Env flag, e.g. `OLMLX_METAL_CAPTURE=/tmp/run.gputrace`.
- Call `mx.metal.start_capture(path)` before generate, `stop_capture()` after.
- Gives true GPU kernel timing via Xcode Instruments. Best for one-off deep dives; not something you'd leave on.

**C. Flash I/O timing** (separate, actually cheap)
- `engine/flash/prefetch.py::PrefetchStats` and `engine/flash/weight_store.py` already do some of this.
- SSD read timing is CPU-side and non-lazy, so it's safe to add without forced syncs.
- This could land independently of A/B.

### Recommendation

Split into: (1) Flash I/O timing as its own small change (low risk, genuinely useful), (2) opt-in `OLMLX_PROFILE` block-level instrumentation with server-log output only, (3) optional Metal-capture hook. Do **not** extend the public `TimingStats` / API response with these fields.

**Comment by dpalmqvist** [comment]
Closing as won't-fix.

No observed bottleneck motivates this. Adding per-op timing (attention/FFN/logits/KV/IO) means inserting `mx.synchronize()` between every operation — that itself perturbs throughput and defeats MLX's lazy-eval pipelining, so the numbers wouldn't reflect real inference anyway. For actual profiling I'd reach for Instruments / Metal System Trace, not bespoke wall-clock counters inside the inference loop.

If a specific bottleneck shows up in real use, I'll instrument that one spot. No need for general scaffolding.
